### PR TITLE
feat: Normalize known DB errors to a MedusaError when possible

### PR DESCRIPTION
**Pull request by sradevski**
Before we would swallow the error and return a generic error to the user. This will provide more information to the caller if it is one of the known errors.

### CODE REVIEW

File: packages/utils/src/dal/mikro-orm/integration-tests/__tests__/mikro-orm-repository.spec.ts [hunk 1]
```diff
@@ -774,5 +784,73 @@ describe("mikroOrmRepository", () => {
...
+    //   })
+
+    //   expect(listedEntities).toHaveLength(2)
+    //   expect(listedEntities[0].entity3.getItems()).toEqual(
+    //     listedEntities[1].entity3.getItems()
+    //   )
+    // })
+  })
+
+  describe.only("error mapping", () => {
```
Comment (adrien2p) [inline packages/utils/src/dal/mikro-orm/integration-tests/__tests__/mikro-orm-repository.spec.ts]: todo: should we remove the only?

**Review by adrien2p — APPROVED** [review]

File: packages/utils/src/dal/mikro-orm/integration-tests/__tests__/mikro-orm-repository.spec.ts (hunk 1 above)
Comment (sradevski) [reply packages/utils/src/dal/mikro-orm/integration-tests/__tests__/mikro-orm-repository.spec.ts]: Definitely 😄

**Review by olivermrbl — APPROVED** [review]
great addition @sradevski!

File: packages/utils/src/dal/mikro-orm/db-error-mapper.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    return null
+  }
+
+  const [keys, values] = detail.match(/\([^\(.]*\)/g) || []
+
+  if (!keys || !values) {
+    return null
+  }
+
+  return {
```
Comment (carlos-r-l-rodrigues) [inline packages/utils/src/dal/mikro-orm/db-error-mapper.ts:72]: Can we also return the name of the constraint (err.constraint)? I believe adding it to the message in parentheses would be helpful.

**Review by carlos-r-l-rodrigues — APPROVED** [review]

File: packages/utils/src/dal/mikro-orm/db-error-mapper.ts (hunk 2 above)
Comment (sradevski) [reply packages/utils/src/dal/mikro-orm/db-error-mapper.ts:72]: @carlos-r-l-rodrigues I think that might be too much of an implementation detail that we leak to the FE. What would be nice though is to have it as part of MedusaError in some form and then we plug that into some error logging/tracing before responding to the end user, but I think we can tackle that later when we integrate better with some tracing library

Comment (carlos-r-l-rodrigues) [reply packages/utils/src/dal/mikro-orm/db-error-mapper.ts:72]: Here is the DAL utils, so I expect the backend to have access to these properties.
it is alright to throw from here directly to the front end, but we need to have access to the details on the backend, even to handle errors differently based on the constraint name.

Comment (sradevski) [reply packages/utils/src/dal/mikro-orm/db-error-mapper.ts:72]: @carlos-r-l-rodrigues yeah definitely, whichi s why I think we should add the constraint on MedusaError as a separate field, but not as part of the message. I suggest we just handle that separately though

**Pull request merged**